pact/Pact/Core/Gas.hs

```diff
@@ -246,15 +231,44 @@ constantGasModel unitPrice gl
   }
 
 freeGasModel :: GasModel b
-freeGasModel = constantGasModel mempty (MilliGasLimit mempty)
+freeGasModel = constantGasModel mempty (MilliGasLimit (MilliGas maxBound)) -- TODO: some tests seem to charge gas even with freeGasModel.
+
+data GasMEnv
+  = GasMEnv
+  { _gasMRef :: IORef MilliGas
+  , _gasMLimit :: MilliGasLimit
+  }
 
-millisPerGas :: Word64
-millisPerGas = 1000
+newtype GasM e a
+  = GasM (ReaderT GasMEnv (ExceptT e IO) a)
+  deriving
+  ( Functor
+  , Applicative
+  , Monad
+  , MonadReader GasMEnv
+  , MonadError e
+  , MonadIO) via (ReaderT GasMEnv (ExceptT e IO))
 
-gasToMilliGas :: Gas -> MilliGas
-gasToMilliGas (Gas n) = MilliGas (n * millisPerGas)
-{-# INLINE gasToMilliGas #-}
+runGasM
+  :: i
+  -> GasMEnv
+  -> GasM (PactError i) a
+  -> IO (Either (PactError i) a)
+runGasM info env (GasM m) = do
+  res <- try $ runExceptT $ runReaderT m env
+  let fakeStackFrame = []
```

Comment:
if we make a version of `runGasM` that works inside of `MonadEval`, we should be able to have a version of this combinator that throws with a stack frame :)

`runGasMEval` maybe?